matmul.hip

```diff
@@ -2330,54 +2336,90 @@ class MmtKernel_StreamK_256t_MSxNS_amdgcn_mfma_f32_16x16x16f16_shared_Kx2
-        // a local accumulator buffer that we can store our partial accumulator
-        // to. That depends on our CU's position in its cooperating group.
+      int cooperating_cu_count = cooperating_cu_end - cooperating_cu_start;
+
+      // Hard case: cooperating with other CUs on this tile.
+      if (cooperating_cu_count != 1) {
         int cu_idx_in_cooperating_group = cu - cooperating_cu_start;
         int local_accumulator_idx =
             cu_idx_in_cooperating_group +
             tile_idx * local_accumulators_count_per_tile(
-                           MNKShape{M_outer, N_outer, K_outer});
+                            MNKShape{M_outer, N_outer, K_outer});
         floatx4_t *local_accumulator = reinterpret_cast<floatx4_t *>(
             local_accumulators_buffer +
             local_accumulator_idx * static_M_tile * static_N_tile);
-        // Perform the MFMA arithmetic, store to our local partial accumulator.
-        mac_loop(A_data, B_data, local_accumulator, K_outer, m_outer, n_outer,
-                 local_iter, local_iter_end);
-        // Now, conditionally on an atomic counter telling us if it's for us to
-        // do, maybe perform the final reduction.
-        maybe_do_final_reduction(C_ptr, M_outer, N_outer, K_outer,
-                                 atomic_counter, local_accumulators_buffer,
-                                 tile_idx, cooperating_cu_start,
-                                 cooperating_cu_end);
+
+        // Store our own accumulators to memory. We don't know yet which CU will
+        // perform the final reduction. If it is not us, then it will need our
+        // local accumulators to have been stored to global memory. If it is us,
+        // then this store is potentially redundant, however:
+        // 1. Something needs to provide memory ordering on the local accumulator
+        //    accesses made by different CUs. At the moment, that thing is the
+        //    same barrier that also determines which CU does the final reduction.
+        //    If we try to potentially save the global store here, we will need
+        //    to add a second fence.
+        // 2. Trying to keep this in registers would mean more register pressure
+        //    and potentially a new constraint on kernel tile size choice.
+        // 3. Likewise if trying to keep that in shared memory.
+        acc_store(acc, local_accumulator);
+
+        atomic_t *atomic_counter = reinterpret_cast<atomic_t *>(
+          aux_bytes + tile_idx * atomics_alignment);
+
+        // Increment the atomic counter and agent-scope acquire-release fence.
+        // This has the effect of ordering the above accumulator store before
+        // the below accumulator loads across CUs ("agent scope").
+        int counter_value = workgroup_atomic_increment(atomic_counter);
+
+        // Determine if we are the CU that should perform the final reduction.
+        if (counter_value == cooperating_cu_count) {
```

Comment:
Not following how this works. If this is false (i.e. we aren't the accumulating CU), then this looks like it still falls through to line 2396 and stores the result to `C_ptr`. That is an obvious race because the accumulating CU could jump in and finish accumulating before we do the false store there.

Reply:
Great catch. Should be addressed by the added commit (separate for review convenience; small).  Also fixed a 3% perf regression.